src/cargo/core/compiler/output_sbom.rs

```diff
@@ -0,0 +1,239 @@
+    version: String,
+    wrapper: Option<PathBuf>,
+    commit_hash: Option<String>,
+    host: String,
+}
+
+impl From<&Rustc> for SbomRustc {
+    fn from(rustc: &Rustc) -> Self {
+        Self {
+            version: rustc.version.to_string(),
+            wrapper: rustc.wrapper.clone(),
+            commit_hash: rustc.commit_hash.clone(),
+            host: rustc.host.to_string(),
+        }
+    }
+}
+
+#[derive(Serialize)]
+struct Sbom {
+    format_version: SbomFormatVersion<1>,
+    package_id: PackageIdSpec,
+    name: String,
+    version: String,
+    source: String,
+    target: SbomTarget,
+    profile: Profile,
+    packages: Vec<SbomPackage>,
+    features: Vec<String>,
+    rustc: SbomRustc,
+}
+
+impl Sbom {
+    pub fn new(unit: &Unit, packages: Vec<SbomPackage>, rustc: SbomRustc) -> Self {
+        let package_id = unit.pkg.summary().package_id().to_spec();
+        let name = unit.pkg.name().to_string();
+        let version = unit.pkg.version().to_string();
+        let source = unit.pkg.package_id().source_id().to_string();
+        let target = (&unit.target).into();
+        let profile = unit.profile.clone();
+        let features = unit.features.iter().map(|f| f.to_string()).collect();
+
+        Self {
+            format_version: SbomFormatVersion,
+            package_id,
+            name,
+            version,
+            source,
+            target,
+            profile,
+            packages,
+            features,
+            rustc,
+        }
+    }
+}
+
+/// Saves a `<artifact>.cargo-sbom.json` file for the given [`Unit`].
+///
+pub fn output_sbom(build_runner: &mut BuildRunner<'_, '_>, unit: &Unit) -> CargoResult<()> {
+    let bcx = build_runner.bcx;
+    let rustc: SbomRustc = bcx.rustc().into();
+
+    let packages = fetch_packages(build_runner, unit);
+
+    for sbom_output_file in build_runner.sbom_output_files(unit)? {
+        let sbom = Sbom::new(unit, packages.clone(), rustc.clone());
+
+        let mut outfile = BufWriter::new(paths::create(sbom_output_file)?);
+        let output = serde_json::to_string(&sbom)?;
+        write!(outfile, "{}", output)?;
+    }
+
+    Ok(())
+}
+
+/// Fetch all dependencies, including transitive ones. A dependency can also appear multiple times
+/// if it's included with different versions.
+fn fetch_packages(build_runner: &mut BuildRunner<'_, '_>, unit: &Unit) -> Vec<SbomPackage> {
+    let unit_graph = &build_runner.bcx.unit_graph;
+    let root_deps = build_runner.unit_deps(unit);
+
+    let mut result: Vec<SbomPackage> = Vec::new();
+    let mut queue: BTreeSet<&UnitDep> = root_deps.iter().collect();
+    let mut visited: BTreeSet<&UnitDep> = BTreeSet::new();
+
+    while let Some(package) = queue.pop_first() {
+        if visited.contains(package) {
+            continue;
+        }
+
+        let build_type = if package.unit.mode.is_run_custom_build() {
+            SbomBuildType::Build
+        } else {
+            SbomBuildType::Normal
+        };
```

Comment:
Could you share the idea of distinguishing the build script execution?

It is a bit odd that the only "crate" dependency a build script execution has is the build script "compilation".

Let me explain it a bit: In Cargo's unit graph there are two kinds of unit for build scripts. One for build script compilation, the other for execution. See this doc for more: <https://doc.rust-lang.org/nightly/nightly-rustc/cargo/core/compiler/custom_build/index.html>.

Yeah we got another similar question: what do we want to track in SBOM?

Reply:
Just re-read the RFC and asked there <https://github.com/rust-lang/rfcs/pull/3553#discussion_r1593283479>.

Reply:
Has this question been resolved?